### PR TITLE
[CBR-179] Bouncing and throttling

### PR DESCRIPTION
**Pull request by parsonsmatt**
## Description

This PR adds throttling capability to the Wallet API as well as the ability to load the configuration from the file.

Note: This requires the `configuration.yaml` file be modified to include the `wallet` section. We should figure out how that affects clients.

## Linked issue

https://iohk.myjetbrains.com/youtrack/issue/CBR-179

## Type of change

- [~] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🛠 New feature (non-breaking change which adds functionality)
- [x] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [~] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [~] 🔨 New or improved tests for existing code
- [~] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

I have not added new tests for this feature. The sort of test we could implement to verify throttling works is to spam the server with a bunch of requests and verify that some of them are `429`. However, the test-suite for the library already does this, so we'd be replicating the work.

Existing tests already catch parsing the config format.

### CODE REVIEW

**Comment by KtorZ** [comment]
Conflicts come from the split recently done between the legacy wallet and the new one. 
I'll try to solve them and get this merged, then possibly issue a PR to cover throttling with the new wallet.

**Comment by KtorZ** [comment]
Okay, I squashed a bunch of stuff together while rebasing and resolving conflicts. Maaaaaaan, that was something. We really need to work on our git practices :cry:  ...

Also, I've made one slight change instead of solving 18 times the same conflict. I've defined a higher abstraction `natMapClient` which can be used to define `hoistClient` and `mapClientErrors`, instead of having to do this plumbing twice in both functions.

```hs
-- | Natural transformation + mapping on a 'WalletClient'
natMapClient
    :: (forall x. m x -> n x)
    -> (forall x. n (Either ClientError x) -> n (Either ClientError x))
    -> WalletClient m
    -> WalletClient n
natMapClient phi f wc = WalletClient
    { getAddressIndexPaginated =
        \x -> f . phi . getAddressIndexPaginated wc x
     -- ... etc

-- | Run the given natural transformation over the 'WalletClient'.
hoistClient :: (forall x. m x -> n x) -> WalletClient m -> WalletClient n
hoistClient phi =
    natMapClient phi id

-- | Generalize a @'WalletClient' 'IO'@ into a @('MonadIO' m) =>
-- 'WalletClient' m@.
liftClient :: MonadIO m => WalletClient IO -> WalletClient m
liftClient = hoistClient liftIO

mapClientErrors :: forall m. Monad m => ResponseErrorHandler m -> WalletClient m -> WalletClient m
mapClientErrors handler =
    natMapClient id overError
```

**Comment by KtorZ** [comment]
@disassembler @cleverca22 
Just notifying you guys about this configuration change:

https://github.com/input-output-hk/cardano-sl/pull/3431/files#diff-61feea0f7a7bc722b74eedc3907be256

This can most probably be left as `null` for now, unless an exchange explicitly requires it. Not sure how this impacts the installer or surrounding scripts.

**Comment by angerman** [comment]
🙄 👀

File: CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,13 @@
   each slot in an epoch) to an epoch/index file pair. Consolidation happens on-the-fly in a
   background process.
 
+- #### Add bouncing and throttling to the API
+  Previously, exchanges could accidentally overload their wallet servers. We
```
Comment (akegalj) [inline CHANGELOG.md]: there was a throttling in v0 per handler https://github.com/input-output-hk/cardano-sl/blob/develop/wallet/src/Pos/Wallet/Web/Methods/Payment.hs#L80

File: lib/configuration.yaml [hunk 2]
```diff
@@ -135,6 +135,21 @@ dev: &dev
...
+    # Configuration options for throttling the API. Items at the top level
+    # apply as a default setting to all requests. A 'null' value indicates that
+    # throttling is disabled.  To enable throttling, change the null value for
+    # the keys that are commented out below.
+    throttle: null
+    # throttle:
+    # # Number of requests per throttling period.
+    #   rate: 200
+    #   # Amount of microseconds in a throttling period.
+    #   period: 1_000_000
```
Comment (akegalj) [inline lib/configuration.yaml]: https://github.com/input-output-hk/cardano-sl/blob/develop/wallet/src/Pos/Wallet/Web/Methods/Payment.hs#L80 note that we have "throttled" 6 seconds here - which is 6 times more then here. I know new data layer is probably more optimized then the old one - but maybe we should measure or check why did we pick 6 in v0. I know it was due to sev1 performance problem

File: lib/src/Pos/Launcher/Configuration.hs [hunk 3]
```diff
@@ -68,6 +73,40 @@ instance FromJSON Configuration where
...
+data ThrottleSettings = ThrottleSettings
+    { tsRate   :: !Word64
+    , tsPeriod :: !Word64
+    , tsBurst  :: !Word64
+    } deriving (Show, Generic)
+
+defaultThrottleSettings :: ThrottleSettings
+defaultThrottleSettings = ThrottleSettings
+    { tsRate = 30
+    , tsPeriod = 1000000
```
Comment (akegalj) [inline lib/src/Pos/Launcher/Configuration.hs]: maybe set to 6sec? see https://github.com/input-output-hk/cardano-sl/blob/develop/wallet/src/Pos/Wallet/Web/Methods/Payment.hs#L80

File: lib/configuration.yaml (hunk 2 above)
Comment (parsonsmatt) [reply lib/configuration.yaml]: Note that this is actually quite a bit faster -- we have `rate` requests per `period` microseconds, so 200 requests permitted per second in this configuration, across the board. It is disabled entirely by default.

More fine-grained throttling can easily be done at the handler level, I think.

Comment (akegalj) [reply lib/configuration.yaml]: > so 200 requests permitted per second in this configuration

yes - we just have to check eventually is our wallet layer up to the task. I am not sure have we run benchmarks on new wallet layer and compared it to the old one

**Comment by akegalj** [comment]
@parsonsmatt 
note that I wanted to do few modification to your PR before it gets merged to develop https://github.com/input-output-hk/cardano-sl/pull/3581

**Pull request merged**